File: frontend/src/api/template.ts
// Typed wrappers for the reusable budget TEMPLATE (plan template UI).
//
// The template is the «set-once, auto-applied to every NEW month» source:
//   - per-category limits   (TemplateItemRead {category_id, limit_cents})
//   - recurring detail lines (TemplateLineRead {id, category_id, kind,
//     amount_cents, day_of_period, title})
//
// It is applied to a new period server-side; the current period's plan is NOT
// touched when the template changes. The template is edited only DIRECTLY
// (ADR-0007 removed the «перезаписать шаблон текущим планом» save-current route).
//
// Endpoints (BACKEND contract /api/v1/template/*):
//   GET    /template/items                 → TemplateItemRead[]
//   PUT    /template/items/{categoryId}     → TemplateItemRead   (upsert limit)
//   GET    /template/lines                  → TemplateLineRead[]
//   POST   /template/lines                  → TemplateLineRead
//   PATCH  /template/lines/{lineId}         → TemplateLineRead
//   DELETE /template/lines/{lineId}         → 204
//
// Wire shapes come straight from the generated contract types (TemplateItem* /
// TemplateLine*), re-exported here under their canonical names.

import { apiFetch } from './client';
import type { components } from './generated/schema';

// ─────────── Wire types (from the generated contract) ───────────

export type TemplateItemRead = components['schemas']['TemplateItemRead'];
export type TemplateItemUpsert = components['schemas']['TemplateItemUpsert'];
export type TemplateLineRead = components['schemas']['TemplateLineRead'];
export type TemplateLineCreate = components['schemas']['TemplateLineCreate'];
export type TemplateLineUpdate = components['schemas']['TemplateLineUpdate'];

// ─────────── In-memory cache (light; invalidated on every mutation) ───────────
//
// The template screen reloads on mount, so a light per-tab cache is enough. Any
// mutation (item upsert / line create-update-delete) clears it so the next read
// re-fetches the authoritative server state.

let itemsCache: TemplateItemRead[] | null = null;
let linesCache: TemplateLineRead[] | null = null;

function invalidateTemplateCache(): void {
  itemsCache = null;
  linesCache = null;
}

// ─────────── Template items (per-category limits) ───────────

/** GET /api/v1/template/items — list of {category_id, limit_cents}. */
export async function getTemplateItems(): Promise<TemplateItemRead[]> {
  if (itemsCache != null) return itemsCache;
  const rows = await apiFetch<TemplateItemRead[]>('/template/items');
  itemsCache = rows;
  return rows;
}

/**
 * PUT /api/v1/template/items/{categoryId} — upsert the template limit for a
 * category. `limitCents` must be ≥ 0 (422 otherwise).
 */
export async function putTemplateItem(
  categoryId: number,
  limitCents: number,
): Promise<TemplateItemRead> {
  const body: TemplateItemUpsert = { limit_cents: limitCents };
  const res = await apiFetch<TemplateItemRead>(
    `/template/items/${categoryId}`,
    { method: 'PUT', body: JSON.stringify(body) },
  );
  invalidateTemplateCache();
  return res;
}

// ─────────── Template lines (recurring detail) ───────────

/** GET /api/v1/template/lines — recurring template lines. */
export async function getTemplateLines(): Promise<TemplateLineRead[]> {
  if (linesCache != null) return linesCache;
  const rows = await apiFetch<TemplateLineRead[]>('/template/lines');
  linesCache = rows;
  return rows;
}

/** POST /api/v1/template/lines — create a recurring template line. */
export async function createTemplateLine(
  payload: TemplateLineCreate,
): Promise<TemplateLineRead> {
  const res = await apiFetch<TemplateLineRead>('/template/lines', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
  invalidateTemplateCache();
  return res;
}

/** PATCH /api/v1/template/lines/{lineId} — partial update. */
export async function patchTemplateLine(
  lineId: number,
  payload: TemplateLineUpdate,
): Promise<TemplateLineRead> {
  const res = await apiFetch<TemplateLineRead>(`/template/lines/${lineId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  });
  invalidateTemplateCache();
  return res;
}

/** DELETE /api/v1/template/lines/{lineId} — hard delete. 204 → success. */
export async function deleteTemplateLine(lineId: number): Promise<void> {
  await apiFetch<void>(`/template/lines/${lineId}`, { method: 'DELETE' });
  invalidateTemplateCache();
}

File: app/api/schemas/template.py
"""Pydantic schemas for v1.1 plan-template endpoints (AGREED §B/§C).

Template = reusable, non-per-period plan. ``items`` carry per-category limits;
``lines`` carry recurring detail rows.
"""

from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


TemplateKindStr = Literal["expense", "income"]


class TemplateItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    category_id: int
    limit_cents: int


class TemplateItemUpsert(BaseModel):
    limit_cents: int = Field(ge=0)


class TemplateLineRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    category_id: int
    title: str
    amount_cents: int
    day_of_period: Optional[int]
    kind: TemplateKindStr


class TemplateLineCreate(BaseModel):
    category_id: int = Field(gt=0)
    title: str = Field(min_length=1, max_length=200)
    amount_cents: int = Field(gt=0)
    day_of_period: Optional[int] = Field(default=None, ge=1, le=31)
    kind: TemplateKindStr


class TemplateLineUpdate(BaseModel):
    category_id: Optional[int] = Field(default=None, gt=0)
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    amount_cents: Optional[int] = Field(default=None, gt=0)
    day_of_period: Optional[int] = Field(default=None, ge=1, le=31)
    kind: Optional[TemplateKindStr] = None


class TemplateRead(BaseModel):
    """Full plan template — per-category limits (items) + recurring lines.

    Response of POST /template/save-current (overwrite-from-current-period).
    """

    items: list[TemplateItemRead]
    lines: list[TemplateLineRead]
